### ISSUE TITLE
Feature request: Add support for parallel install of snaps

### ISSUE THREAD

**Issue by hemanthnakkina** [canonical/operator-libs-linux#134]
This is a feature request to add support for parallel install of snap.

This will be a minor change to retry getting snap info removing suffix instance key while getting snap information [2].

[1] https://snapcraft.io/docs/parallel-installs
[2] https://github.com/canonical/operator-libs-linux/blob/a1aaa35135220d564f7a63854388b38752de09c4/lib/charms/operator_libs_linux/v2/snap.py#L907

**Comment by benhoyt** [comment]
Thanks @hemanthnakkina. Can you please be a bit more specific and detailed about what you're thinking this feature/API would look like?

**Comment by hemanthnakkina** [comment]
@benhoyt There are no API changes required for the lib to install a snap with <snap name>_<instance key> (the format used for snap in parallel install). However the lib currently breaks in maintaining SnapCache as snapcache tries to get information about the snap using the snap_name passed. Instead logic can be added in SnapCache to check the possibility of parallel install snap name format and retry getting snap info using actual snap name.

Changes might look like this
```
diff --git a/lib/charms/operator_libs_linux/v2/snap.py b/lib/charms/operator_libs_linux/v2/snap.py
index 9d09a78..979b593 100644
--- a/lib/charms/operator_libs_linux/v2/snap.py
+++ b/lib/charms/operator_libs_linux/v2/snap.py
@@ -904,10 +904,16 @@ class SnapCache(Mapping):
         Args:
             name: a string representing the name of the snap
         """
-        info = self._snap_client.get_snap_information(name)
+        try:
+            info = self._snap_client.get_snap_information(name)
+            name = info["name"]
+        except SnapAPIError:
+            # Check if the snap name is parallel install with instance key
+            snap_name = name.rsplit('_')[0]
+            info = self._snap_client.get_snap_information(snap_name)
 
         return Snap(
-            name=info["name"],
+            name=name,
             state=SnapState.Available,
             channel=info["channel"],
             revision=info["revision"],
```

There are no new parameters/functions exposed by the lib and so the current users are not impacted.
The change will be compatible with existing version.

What do you think?

**Comment by benhoyt** [comment]
Hi @hemanthnakkina, we'll do a bit more research on this and try to fix this early in the 25.04 cycle.

**Comment by james-garner-canonical** [comment]
My initial concern here is that, as proposed, this retry logic could potentially complicate tracebacks for non-parallel users of the lib. I'm also concerned that there might be unforeseen issues related to handling parallel snaps lurking, that will require additional special casing as they are discovered.

I also suspect that parallel installs would 'just work' if we went with a Jubilant-style rewrite -- a thin wrapper over the snap cli instead of the lib's current cache oriented interface. I propose that we defer work on parallel snap support until such a rewrite, and take parallel snap support as a motivating case for it. Thoughts?

**Comment by benhoyt** [comment]
Yeah, that (focusing our efforts on a subprocess.run-style rewrite) sounds better to me too.
